pil-analyzer/src/structural_checks.rs

```diff
@@ -0,0 +1,117 @@
+use itertools::Itertools;
+use std::collections::{HashMap, HashSet};
+
+use powdr_ast::{
+    analyzed::{Expression, FunctionValueDefinition, PolynomialReference, Reference, Symbol},
+    parsed::{StructExpression, TypeDeclaration},
+};
+use powdr_parser_util::{Error, SourceRef};
+
+/// Verifies that all struct instantiations match their corresponding declarations
+/// (existence of field names, completeness) and ensures that both are correct.
+pub fn check_structs_fields<'a>(
+    structs_exprs: impl Iterator<Item = &'a Expression>,
+    definitions: &HashMap<String, (Symbol, Option<FunctionValueDefinition>)>,
+) -> Result<(), Vec<Error>> {
+    let mut errors = Vec::new();
+
+    for expr in structs_exprs {
+        let Expression::StructExpression(
+            source,
+            StructExpression {
+                name: Reference::Poly(PolynomialReference { name, .. }),
+                fields,
+            },
+        ) = expr
+        else {
+            unreachable!()
```

Comment:
To be honest, instead of this I would prefer just filtering the expressions here instead of in the analyzer, but you can do this in a separate pr.